메인 화면에서 파일을 드롭다운하여 업로드할 수 있는 기능 구현

always-awake opened the pull request:
# 메인 화면에서 파일을 드롭다운하여 업로드할 수 있는 기능 구현

## 해당 이슈 📎

- [Epic#11] 메인 화면에서 슬라이드를 드롭다운 할 수 있다. (#48)

## 변경 사항 🛠
- 메인화면에서 파일은 드롭다운하여 이미지 컨버터 서버에 파일을 전송할 수 있다.
- 이미지 컨버터 서버의 처리 결과가 'ok'가 아닐 경우, ErrorModal을 렌더링 할 수 있도록 구현하였다.
- 파일이 DropZone에 drop될 때부터 채널이 생성이되고 특정 채널 페이지로 리다이렉트 될 때까지 LoadingModal을 렌더링 할 수 있도록 구현하였다.
- Spinner common 컴포넌트 구현

## 테스트 ✨
> 없음

## 리뷰어 참고 사항 🙋‍♀️
- 각 컴포넌트의 파일명과 파일 구조의 적절성
- ErrorModal, LoadinModal, Spinner 등의 컴포넌트의 역할배분/구조화가 잘되어 있는지 궁금합니다.

always-awake committed:
[ BUGFIX ] tempFiles 폴더에 업로드된 파일명이 undefined.pdf 로 표시되는 오류 해결 (#23)

Why
- remove 미들웨어가 실행될 때, 의도치 않게 tempFiles 내 모든 pdf 파일을 삭제하게 된다.

What
- createFormData에 넘겨주는 인자의 순서를 channelId, file로 변경하였다.

always-awake committed:
[ FEATURE ] 파일 드롭다운 시,  파일이 image-converter 서버로 전송되는 로직 구현 (#48)

Why
- 사용자는 파일을 화면에 드롭다운하여, 자신의 파일을 업로드할 수 있어야 한다.
- 사용자가 업로드한 파일은 이미지 컨버터 서버로 전송되어야 한다.

What
- onDrop 이벤트를 처리하는 handleDrop 함수를 구현하였다.
- handleDrop 함수는 drop됬을 때, event 객체의 dataTransfer.files의 첫번째 file을 이미지 컨버터 서버에 전송한다.
- 이미지 컨버터 서버의 응답이 'ok'일 경우, useCreateChannel 훅을 이용해 채널을 생성한다.
- DropModal 컴포넌트명을 더욱 직관적인 DropZone으로 수정하였다.

always-awake committed:
[ FEATURE ] 로딩 시, 렌더링할 스피너 컴포넌트 구현 (#48)

Why
- 로딩 중인 상황을 사용자가 알 수 있도록 해야 한다.

What
- material ui 의 CircularProgress 컴포넌트를 사용하였다.
- S.SpinnerWrapper로 CircularProgress 컴포넌트를 감싸서, z-index의 값을 999로 설정하였다.

always-awake committed:
[ FEATURE ] 로딩 시, 렌더링할 LoadingText 컴포넌트 구현 (#48)

Why
- 로딩 시, 유저에게 현재 진행중인(로딩중인) 작업에 대한 정보를 전달해야 한다.

What
- message props를 전달받으면, 해당 message 문자열을 보여주는 LoadingText 컴포넌트 구현하였다.
- LoadingText 컴포넌트는 스타일 컴포넌트를 이용하여 스타일링하였다.

always-awake committed:
[ FEATURE ] 로딩 시, 렌더링할 LoadingModal 컴포넌트 구현 (#48)

Why
- 로딩 시, 데이터 처리가 완료될 때까지 유저에게 로딩화면을 보여주어야 한다.

What
- LoadingModal 컴포넌트가 message prop을 받아 LoadingText에 전달하도록 구현하였다.
- LoadingModal 컴포넌트가 Spinner 컴포넌트와 LoadingText 컴포넌트를 함께 렌더링하도록 구현하였다.
- LoadingModal 컴포넌트의 스타일링은 스타일 컴포넌트를 이용하여 구현하였다.

always-awake committed:
[ CHORE ] 전역으로 사용할 theme 객체 내 primary의 light 색깔 수정

Why
- 전역으로 사용할 색깔이 잘못된 색깔로 설정되어 있기 때문에 이를 수정해야 한다.

What
- 기존에 #343a40 로 설정되어 있던 것을 #FAB005 로 수정하였다.
- styles/index.js 의 소스파일 마지막 줄바꿈 lint 에러를 수정하였다.

always-awake committed:
[ FEATURE ] 에러 발생시, 에러 메시지를 렌더링하는 ErrorText 컴포넌트 구현 (#48)

Why
- 에러가 발생하면, 유저에게 에러에 대한 정보를 전달해야 한다.

What
- message props를 전달받으면, 해당 message 문자열을 보여주는 ErrorText 컴포넌트를 구현하였다.

always-awake committed:
[ FEATURE ] 에러 발생시, 렌더링할 ErrorEmoji 컴포넌트 구현 (#48)

Why
- 에러가 발생하면, 에러가 발생했음을 쉽게 알 수 있는 이모지를 보여주어야 한다.
What
- 후라이펜 모양의 이모지를 렌더링하는 ErrorEmogi 컴포넌트를 구현하였다.
- ErrorEmoji는 스타일 컴포넌트를 이용하여 스타일링 하였다.

always-awake committed:
[ FEATURE ] 에러가 발생 시, 에러에 대한 정보를 전달하는 ErrorModal 컴포넌트 구현 (#48)

Why
- 에러가 발생하면, 사용자에게 에러에 대한 정보를 전달해야 한다.
- 에러가 발생하면, 사용자가 한눈에 에러임을 알 수 있도록 해야 한다.

What
- message props를 전달받아 ErrorText 컴포넌트에 전달하는 ErrorModal 을 구현하였다.
- ErrorEmoji 컴포넌트와 ErrorText를 렌더링하는 ErrorModal을 구현하였다.

always-awake committed:
[ FEATURE ] 파일을 drop 하였을 때, 사용자 파일을 이미지 컨버터 서버로 전송하는 로직 구현 (#48)

Why
- 파일을 drop하는 이벤트가 발생하면, 이벤트 객체 내 파일을 이미지 컨버터 서버에 업로드해야 한다.
- 이미지 컨버터 서버에 업로드 성공 여부에 따라 다른 처리를 하기 위해 분기 처리해야 한다.

What
- drop 이벤트 객체에서 사용자 file과 channelId를 담고 있는 formData 객체를 생성해준다.
- 생성된 formData를 이미지 컨버터 서버로 전송한다.
- 이미지 컨버터 서버의 응답이 'ok'라면, channelId를 이용해 채널을 생성해주는 로직이 실행되도록 처리한다.
- 'ok'외의 응답이 온다면, setIsError 을 이용해 isError의 값을 true로 변경시켜준다.

always-awake committed:
[ FEATURE ] 파일을 드롭다운하면, 해당 파일을 업로드하는 DropZone 컴포넌트 구현 (#48)

Why
- dropy 핵심 UX인 Drag & Drop 기능을 구현해야 한다.
- 사용자가 Drag & Drop 기능으로 쉽게 사용자의 파일을 업로드할 수 있어야 한다.

What
- 파일을 drag 하고 있을 때 발생하는 이벤트를 처리하는 handleDragOver 함수를 구현하였다.
- 파일을 drag하고 있지 않을 때 발생하는 이벤트를 처리하는 handleDragLeave 함수를 구현하였다.
- isError 값에 따라 ErrorModal의 렌더링 여부가 결정되도록 구현하였다.
- isError 값이 false일 경우, isLoading 값에 따라 LoadingModal의 렌더링 여부가 결정되도록 구현하였다.

always-awake committed:
[ CHORE ] 스타일 컴포넌트들의 타입을 명시해주도록 수정

Why
- 코드의 유지보수를 위해 스타일 컴포넌트의 타입을 명시해주어야 한다.

What
- ErrorModal, LoadingModal, Spinner 등의 스타일 컴포넌트에 타입을 명시해주는 코드를 추가하였다.

always-awake committed:
[ IMPROVE ] DropEmoji 컴포넌트가 props로 받을 emoji를 렌더링할 수 있도록 수정 (#48)

Why
- 고정된 이모지만을 렌더링하지 않고, 매번 다른 이모지를 렌더링하기 위함

What
- DropEmoji는 props내 emoji를 전달받아 렌더링하도록 수정하였다.
- 이모지의 크기를 작게 수정하였다.

always-awake committed:
[ IMPROVE ] fontColor, message에 따라 다른 스타일의 컴포넌트를 렌더링할 수 있도록 수정 (#48)

Why
- DropText 컴포넌트가 고정된 색과 message만을 렌더링하지 않도록 하기 위함

What
- DropText는 props 내 fontColor, message를 전달받아 가변적으로 렌더링하도록 수정하였다.
- 스타일 컴포넌트에서 fontColor 값에 따라 다른 color 값을 설정하도록 분기 처리하였다.

always-awake committed:
[ UTIL ] 자주 사용하는 main color, light-main color를 리턴해주는 유틸함수 구현

Why
- 컴포넌트 스타일 전반적으로 자주 사용되는 색정보를 바로 불러올 수 있도록 저리할 필요가 있다.

What
- theme 객체를 이용하여 메인 색을 조회하는 colorMain 유틸 함수 구현
- theme 객체를 이용하여 light 메인 색을 조회하는 colorLightMain 유틸 함수 구현

always-awake committed:
[ IMPROVE ] 처리 결과에 따라 로딩화면과 에러 화면을 조건부로 렌더링 할 수 있도록 수정 (#48)

Why
- DropInput 컴포넌트를 이용하여 파일을 업로드할 경우, 처리 결과에 따라 사용자에게 로딩/에러 정보를 제공해야 한다.
What
- useState 리액트 훅을 이용하여 isLoading, isError 상태 값을 관리하였다.
- isLoading, isError 값에 따라 ErrorModal 또는 LoadingModal을 조건부 렌더링 하도록 구현하였다.

always-awake committed:
[ STYLE ] 매번 사용자가 파일을 드래그할 때마 다 다른 DropEmoji와 DropText가 렌더링 되도록 수정 (#48)

Why
- 파일을 drag 했을 때, 매번 고정된 이모지와 텍스트가 아닌 가변 적인 이모지와 텍스트를 보여주어야 한다.

What
- DropZone 컴포넌트 내에 파일 drag가 시작되었을 때, 이벤트를 처리하는 handleDragEnter 함수를 구현하였다.
- 특정 리스트의 아이템을 렌덤으로 리턴해주는 getRandomItemOfList 함수를 구현하였다.
- getRandomItemOfList 함수를 이용하여 emojiList 내에 있는 이모 지들을 랜덤하게 dropZoneEmoji 로 설정하였다.

aereeeee commented on web/src/components/common/ErrorModal/ErrorEmoji/style.jsx:
```diff
@@ -0,0 +1,9 @@
+import { px } from '@/styles/themeUtil';
+
+export default {
+  ErrorEmoji: styled.div.attrs({ type: 'div' })`
```
`attrs({ type: 'div' })`은 왜 붙이신건가요?🤔

always-awake replied:
스타일 컴포넌트도 타입을 체크해주는 것이 좋을 것 같아서요!
(저희,, 컨벤션 아니었나용?,,)

always-awake replied:
앗,, 제 실수 입니당 😭 수정하였습니다!

aereeeee commented on web/src/components/common/ErrorModal/ErrorModal.jsx:
```diff
@@ -0,0 +1,24 @@
+  message: PropTypes.string.isRequired,
+};
+
+export default ErrorModal;
```
에러모달을 공통으로 사용할수있게 만들어주셔서 고맙습니당~😘

aereeeee commented on web/src/components/common/ErrorModal/style.jsx:
```diff
@@ -0,0 +1,24 @@
+    align-items: center;
+    width: 100%;
+    height: 100%;
+    background-color: white
```
white컬러도 theme.palette.common.white에서 사용해주면 더 좋을것 같아요!
나중에 white칼라의 톤이 전체적으로 바뀌는 경우를 생각해서요!

always-awake replied:
반영했습니당

aereeeee commented on web/src/components/main/DropZone/DropZone.jsx:
```diff
@@ -0,0 +1,86 @@
+  const handleDragEnter = (event) => {
+    event.preventDefault();
+
+    const emojiList = ['😀', '😃', '🤣', '😍', '😘', '🥳', '😻', '😽', '😇', '🦄', '🤓', '😎', '🤩'];
```
이런 부분이나 default에러메세지 같은 것은 constants로 빼면 좋을 것 같아요! 
나중에 관리하기 쉬울것 같습니다.

always-awake replied:
오 그렇게 수정하는게 좋을것 같습니당 ㅎㅎ 반영했습니다! 👌

aereeeee commented on web/src/components/main/DropZone/DropZone.jsx:
```diff
@@ -0,0 +1,86 @@
+const DropZone = () => {
+  const { mutate, data } = useCreateChannel();
+  const [isLoading, setIsLoading] = useState(false);
+  const [isError, setIsError] = useState(false);
```
위의 DropInput컴포넌트에서도 isLoading과 isError state가 사용되고 있고 두 상태 모두 같은 이벤트를 기반으로 발생하는 것이니 전역으로 관리하는게 좋을것 같아요 context API와 useReducer를 이용하여 간단하게 구현할 수 있을 것 같습니다!
[useReducer로 로딩, 에러상태 관리 예시](https://medium.com/free-code-camp/hooked-on-hooks-how-to-use-reacts-usereducer-2fe8f486b963)

always-awake replied:
와우! 너무너무 좋은 리뷰인것 같습니다 꼭 수정하여 반영하겠습니당 🙆‍♀️

aereeeee commented on web/src/styles/themeUtil.js:
```diff
@@ -1,10 +1,13 @@
 const colorYellow = (val) => ({ theme }) => theme.palette.dropyYellow[val];
-
+const colorMain = () => ({ theme }) => theme.palette.primary.main;
+const colorLightMain = () => ({ theme }) => theme.palette.primary.light;
```
colorMain과 colorLightMain을 하나로 합치면 더 좋을 것 같아요! primary중 인자로 main인지 light인지 넣어주는 방식이 어떨까요? 😁

aereeeee commented on web/src/utils/random.js:
```diff
@@ -0,0 +1,8 @@
+  return targetList[randomIndex];
+};
+
+export default getRandomItemOfList;
```
랜덤으로 이모지를 선택하는 유틸을 만드셨군요!! 짱입니다😙

devjaewon replied:
저도 이 유틸 아주 좋다고 생각합니다!

happydhKim replied:
짱이에용🤭

always-awake replied:
감사합니다잉 😎

aereeeee reviewed (commented):
컴포넌트를 많이 만드셨네요.. 드롭다운했을때 변하는 모습을 보고 감동받았습니다😭 우리 서비스의 핵심이 구현되었네요!!!:-)

happydhKim reviewed (approved):
와우! 코드가 굉장히 깔끔하네요! 
스타일에서 px 같은 것들은 팀회의 때 정해서 어서 같이 수정해요!!
같은 팀이여서 행복해요 🥺

devjaewon commented on web/src/components/common/ErrorModal/ErrorModal.jsx:
```diff
@@ -0,0 +1,24 @@
+  const { message } = props;
+
+  return (
+    <S.ErrorModalWrapper>
```
저도 아직 특별한 컨벤션 없이 사용하고 있는데, `ErrorModal` 내부에 `ErrorModalWrapper` 라는 컴포넌트가 있는 것이 조금 부자연스럽다는 느낌이 듭니다.

다 같이 컨벤션을 "컴포넌트 이름 = 최상위 스타일 컴포넌트 이름" 으로 맞추는 것이 어떨까요?

always-awake replied:
좋은 리뷰 감사합니다! 수정하였습니당

always-awake committed:
[ REFACTOR ] 코드 리뷰 사항을 반영하여 수정

Why
- 리뷰어들에게 리뷰 받은 사항 중 수정이 필요하다고 생각되는 로직을 수정 해야 한다.

What
- 스타일 컴포넌트 내 불필요한 타입 체크 부분을 삭제해준다.
- 컴포넌트에서 Wrapper 로 감싸주어 렌더링하는 부분을 수정하였다.
- 메시지, 이모지 리스트와 같은 상수들은 따로 constants/main.js에 따로 정의해주었다.

always-awake committed:
Merge branch 'develop' into feature/issue48
